sentry/src/main/java/io/sentry/Hub.java

```diff
@@ -566,20 +565,20 @@ public void flush(long timeoutMillis) {
       final @Nullable CustomSamplingContext customSamplingContext) {
     Objects.requireNonNull(transactionContexts, "transactionContexts is required");
 
-    final SamplingContext samplingContext =
-        new SamplingContext(transactionContexts, customSamplingContext);
-    boolean samplingDecision = tracingSampler.sample(samplingContext);
-    transactionContexts.setSampled(samplingDecision);
-
     ITransaction transaction;
     if (!isEnabled()) {
       options
           .getLogger()
           .log(
               SentryLevel.WARNING,
               "Instance is disabled and this 'startTransaction' returns a no-op.");
-      transaction = new NoOpTransaction();
+      transaction = NoOpTransaction.getInstance();
     } else {
+      final SamplingContext samplingContext =
+          new SamplingContext(transactionContexts, customSamplingContext);
+      boolean samplingDecision = tracingSampler.sample(samplingContext);
+      transactionContexts.setSampled(samplingDecision);
```

Comment:
M: Could we even take this decision in the ctor now and hide this setter? THat's until we decide on the right approach to have deferred sampling?

Reply:
What would be the benefit of that? Too many options in constructor means many constructors.

Reply:
See https://github.com/getsentry/develop/issues/243

Reply:
The problem is that changing sampling mid-flight means you send out some sentry-trace with one decision then others with another which will result in broken traces.

Reply:
@bruno-garcia this is apparently allowed by design: https://github.com/getsentry/develop/issues/243#issuecomment-760971206

> What you're asking for is not possible with head-based sampling.
You can, however, overwrite the sampling decision for the current transaction at any point before it goes to Sentry (that is, you can drop the transaction), but you cannot expect that it will affect peer services, because at that point sampling decisions were already propagated via HTTP headers.

If I understand correctly, it's possible to sample out a transaction on creation and also at any point before it's sent to Sentry. In case of the latter, it will not affect peer services.

Reply:
Once we start receiving reports of broken traces, it'll be a nightmare to debug why. We'll play a blame game where we'll say the user must be changing sampling, and users will say we'll no propagating properly, or dropping transactions on the Transport/Relay/else.

@rhcarvalho Am I overreacting here?

Reply:
> it's possible to sample out a transaction on creation and also at any point before it's sent to Sentry

@Tyrrrz 
While it is _technically possible_, I believe that's not how it was envisioned to be used.
Sampling should be done "behind the scenes" when starting the transaction. In Python, JS, etc, there's no stopping anyone from mutating an object later on, though.

The proper way to influence sampling is with an explicit traces sampler.

After some back-and-forth with the Go implementation, I let go of making everything "super protected" to stay inline with how the error reporting part of the SDK works, it is rather liberal.

I'll let ya'll think what design makes more sense for Java.

> Am I overreacting here?

@bruno
Maybe :P